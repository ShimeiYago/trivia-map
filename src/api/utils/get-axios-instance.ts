import { axiosMockError } from 'api/mock/error-response';
import axios, {
  AxiosAdapter,
  AxiosError,
  AxiosInstance,
  AxiosRequestConfig,
  AxiosResponse,
} from 'axios';

/**
 * This returns the AxiosInstance for API request.
 * Axios request returns mock data depending on environment variable "REACT_APP_MOCK".
 * If env "REACT_APP_MOCK" is "normal", axios returns mock response data.
 * And if env "REACT_APP_MOCK" is "error", axios throw mock error.
 *
 * @param additionalConfig - The AxiosRequestConfig additionally set to default
 * @param mockData - The expected mock response data
 * @returns The AxiosInstance for API request
 */

export function getAxiosInstance(
  additionalConfig: AxiosRequestConfig = {},
  mockData?: unknown,
): AxiosInstance {
  const axiosRequestConfig: AxiosRequestConfig = {
    ...defaultConfig,
    ...additionalConfig,
  };

  if (mockData) {
    switch (process.env.REACT_APP_MOCK) {
      case 'normal':
        axiosRequestConfig.adapter = getMockAdapter(mockData, 'normal');
        break;
      case 'error':
        axiosRequestConfig.adapter = getMockAdapter(mockData, 'error');
        break;
    }
  }

  return axios.create(axiosRequestConfig);
}

const defaultConfig: AxiosRequestConfig = {
  timeout: 3000,
};

function getMockAdapter(
  mockData: unknown,
  mode: 'normal' | 'error',
): AxiosAdapter {
  return async () => {
    await sleep(Number(process.env.REACT_APP_MOCK_DELAY_TIME));

    return new Promise((resolve, reject) => {
      switch (mode) {
        case 'normal':
          resolve({ data: mockData } as AxiosResponse);
          break;
        case 'error':
          reject(axiosMockError as AxiosError);
          break;
      }
    });
  };
}

async function sleep(time: number) {
  await new Promise((resolve) => setTimeout(resolve, time));
}
