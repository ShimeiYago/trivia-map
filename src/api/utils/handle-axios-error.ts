import { AxiosError } from 'axios';

export function handleAxiosError(axiosError: AxiosError): ApiError {
  if (axiosError.response) {
    // when API is active but response includes error
    const error: ApiError = {
      status: axiosError.response.status,
      data: axiosError.response.data,
      errorMsg: axiosError.message,
    };
    return error;
  } else {
    // when API is not active
    const error: ApiError = {
      status: 500,
      data: {},
      errorMsg: axiosError.message,
    };
    return error;
  }
}

export type ApiError = {
  status: number;
  data: unknown;
  errorMsg: string;
};
