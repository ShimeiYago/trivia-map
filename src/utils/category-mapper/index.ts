import { CATEGORIES } from '../../constant';

export function categoryMapper(categoryId: number) {
  for (const category of CATEGORIES) {
    if (category.categoryId === categoryId) {
      return category.categoryName;
    }
  }

  return CATEGORIES[CATEGORIES.length - 1].categoryName;
}
