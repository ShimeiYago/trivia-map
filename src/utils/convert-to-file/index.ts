import { SelializedImageFile } from '../../types/selialized-image-file';

export async function convertToFile(selializedFile: SelializedImageFile) {
  const blob = await (await fetch(selializedFile.dataUrl)).blob();
  return new File([blob], selializedFile.fileName);
}
