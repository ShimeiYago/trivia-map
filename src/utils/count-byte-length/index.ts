/* istanbul ignore file */

export function countByteLength(str: string) {
  let r = 0;
  for (let i = 0; i < str.length; i++) {
    const c = str.charCodeAt(i);
    // Shift_JIS: 0x0 ～ 0x80, 0xa0 , 0xa1 ～ 0xdf , 0xfd ～ 0xff
    // Unicode : 0x0 ～ 0x80, 0xf8f0, 0xff61 ～ 0xff9f, 0xf8f1 ～ 0xf8f3
    if (
      (c >= 0x0 && c < 0x81) ||
      c == 0xf8f0 ||
      (c >= 0xff61 && c < 0xffa0) ||
      (c >= 0xf8f1 && c < 0xf8f4)
    ) {
      r += 1;
    } else {
      r += 2;
    }
  }
  return r;
}
