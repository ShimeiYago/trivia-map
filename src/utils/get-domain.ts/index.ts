export function getDomain(window: Window) {
  const uri = new URL(window.location.href);
  return uri.origin;
}
