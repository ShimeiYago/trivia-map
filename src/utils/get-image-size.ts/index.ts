/* istanbul ignore file */

export async function getImageSize(src: string): Promise<Size> {
  // convert object url to image
  const img = new Image();
  await new Promise((resolve, reject) => {
    img.onload = () => resolve(img);
    img.onerror = (e) => reject(e);
    img.src = src;
  });

  return {
    width: img.width,
    height: img.height,
    aspect: img.height / img.width,
  };
}

type Size = {
  width: number;
  height: number;
  aspect: number;
};
