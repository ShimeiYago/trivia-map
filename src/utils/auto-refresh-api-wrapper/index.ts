import { refreshToken } from 'api/auths-api/refresh-token';
import { ApiError } from 'api/utils/handle-axios-error';

export async function autoRefreshApiWrapper<Response>(
  api: () => Promise<Response>,
) {
  let apiError: ApiError<unknown>;

  try {
    return await api();
  } catch (error) {
    apiError = error as ApiError<unknown>;

    if (apiError.status !== 401) {
      throw apiError;
    }
  }

  try {
    await refreshToken();
  } catch (error) {
    throw apiError;
  }

  return await api();
}
