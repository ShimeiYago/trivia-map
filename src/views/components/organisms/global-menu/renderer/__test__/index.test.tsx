import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Renderer, Props, State } from '..';
import * as SleepModule from 'utils/sleep';

const basicProps: Props = {
  topBarPosition: 'fixed',
  children: 'contents',
  openAuthFormModal: false,
  loggedOutSuccessfully: false,
  pathName: '/',
  toggleAuthFormModal: jest.fn(),
  logout: jest.fn(),
  isMobile: false,
  isFormEditting: false,
};

let wrapper: ShallowWrapper<Props, State, Renderer>;

jest.spyOn(SleepModule, 'sleep');

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('basic', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('new page', () => {
    wrapper.setProps({ pathName: '/map/new', isFormEditting: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('edit page', () => {
    wrapper.setProps({ pathName: '/map/edit/1' });
    expect(wrapper).toMatchSnapshot();
  });

  it('static position', () => {
    wrapper.setProps({ topBarPosition: 'static' });
    expect(wrapper).toMatchSnapshot();
  });

  it('permanent left navi', () => {
    wrapper.setProps({ permanentLeftNavi: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('with userInfo', () => {
    wrapper.setProps({
      userInfo: {
        email: 'xxx@example.com',
        userId: 1,
        nickname: 'Axel',
        icon: 'https://...',
        url: 'https://...',
        isSocialAccount: false,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('redirect to top page', () => {
    wrapper.setState({
      redirectToTop: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('with local navi', () => {
    wrapper.setProps({
      localBackNavi: {
        text: 'text',
        link: 'https://...',
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('with local navi & mobile', () => {
    wrapper.setProps({
      localBackNavi: {
        text: 'text',
        link: 'https://...',
      },
      isMobile: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('map page', () => {
    wrapper.setProps({
      mapPage: true,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('long name with mobile', () => {
    wrapper.setProps({
      userInfo: {
        email: 'xxx@example.com',
        userId: 1,
        nickname: 'xxxxxxxxxxxxxxx',
        icon: 'https://...',
        url: 'https://...',
        isSocialAccount: false,
      },
      isMobile: true,
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('componentDidUpdate', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should set authMenuAnchorEl null when loggedOutSuccessfully is changed to true', () => {
    wrapper.setProps({
      loggedOutSuccessfully: true,
    });

    const instance = wrapper.instance();
    instance['componentDidUpdate']({ loggedOutSuccessfully: false } as Props, {} as State);
    expect(instance.state.authMenuAnchorEl).toBe(null);
  });

  it('should set redirectToTop false when redirectToTop is changed to true', () => {
    wrapper.setState({
      redirectToTop: true,
    });

    const instance = wrapper.instance();
    instance['componentDidUpdate']({} as Props, { redirectToTop: false } as State);
    expect(instance.state.redirectToTop).toBeFalsy();
  });
});

describe('toggleLeftMenu', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should open left navi', () => {
    const instance = wrapper.instance();
    instance['toggleLeftMenu'](true)();
    expect(instance.state.openLeftNavi).toBeTruthy();
  });
});

describe('handleClickAuthMenu', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should set anchor element', () => {
    const event = {
      currentTarget: null,
    } as unknown as React.MouseEvent<HTMLElement>;
    const instance = wrapper.instance();
    instance['handleClickAuthMenu'](event);
    expect(instance.state.authMenuAnchorEl).toBe(null);
  });
});

describe('handleCloseAuthPopover', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should reset handleCloseAuthPopover', () => {
    const instance = wrapper.instance();
    instance['handleCloseAuthPopover']();
    expect(instance.state.authMenuAnchorEl).toBe(null);
  });
});

describe('toggleAuthModal', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should open auth modal', () => {
    const instance = wrapper.instance();
    instance['toggleAuthModal'](true, 'login')();
    expect(instance.props.toggleAuthFormModal).toBeCalled();
  });
});

describe('handleLoginSucceed', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should call toggleAuthFormModal', async () => {
    const instance = wrapper.instance();
    await instance['handleLoginSucceed']();
    expect(instance.props.toggleAuthFormModal).toBeCalled();
  });
});

describe('handleClickLogout', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('should call logout', async () => {
    const instance = wrapper.instance();
    await instance['handleClickLogout']();
    expect(instance.props.logout).toBeCalled();
  });
});
