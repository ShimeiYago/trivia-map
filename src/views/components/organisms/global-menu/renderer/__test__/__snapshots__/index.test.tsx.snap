// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests basic 1`] = `
<Fragment>
  <ForwardRef(AppBar)
    position="fixed"
    sx={
      Object {
        "borderBottom": "2px solid rgba(255, 255, 255, 0.7)",
      }
    }
  >
    <ForwardRef(Toolbar)>
      <ForwardRef(IconButton)
        aria-label="open drawer"
        color="inherit"
        edge="start"
        onClick={[Function]}
        sx={
          Object {
            "mr": 2,
          }
        }
      >
        <Memo(ForwardRef(MenuIcon)) />
      </ForwardRef(IconButton)>
      <ForwardRef(Typography)
        component="div"
        sx={
          Object {
            "flexGrow": 1,
          }
        }
        variant="h6"
      >
        Persistent drawer
      </ForwardRef(Typography)>
      <ForwardRef(IconButton)
        aria-controls="menu-appbar"
        aria-haspopup="true"
        aria-label="account of current user"
        color="inherit"
        size="large"
      >
        <Memo(ForwardRef(AccountCircleIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(Toolbar)>
  </ForwardRef(AppBar)>
  <ForwardRef(SwipeableDrawer)
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
  >
    <ForwardRef(Box)
      sx={
        Object {
          "height": "100%",
          "width": 250,
        }
      }
    >
      <ForwardRef(List)>
        <ForwardRef(ListItem)
          button={true}
          key="Inbox"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Inbox"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Starred"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MailIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Starred"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Send email"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Send email"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Drafts"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MailIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Drafts"
          />
        </ForwardRef(ListItem)>
      </ForwardRef(List)>
      <ForwardRef(Divider) />
      <ForwardRef(List)>
        <ForwardRef(ListItem)
          button={true}
          key="All mail"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="All mail"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Trash"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MailIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Trash"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Spam"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Spam"
          />
        </ForwardRef(ListItem)>
      </ForwardRef(List)>
    </ForwardRef(Box)>
  </ForwardRef(SwipeableDrawer)>
</Fragment>
`;

exports[`Shallow Snapshot Tests static position 1`] = `
<Fragment>
  <ForwardRef(AppBar)
    position="static"
    sx={
      Object {
        "borderBottom": "2px solid rgba(255, 255, 255, 0.7)",
      }
    }
  >
    <ForwardRef(Toolbar)>
      <ForwardRef(IconButton)
        aria-label="open drawer"
        color="inherit"
        edge="start"
        onClick={[Function]}
        sx={
          Object {
            "mr": 2,
          }
        }
      >
        <Memo(ForwardRef(MenuIcon)) />
      </ForwardRef(IconButton)>
      <ForwardRef(Typography)
        component="div"
        sx={
          Object {
            "flexGrow": 1,
          }
        }
        variant="h6"
      >
        Persistent drawer
      </ForwardRef(Typography)>
      <ForwardRef(IconButton)
        aria-controls="menu-appbar"
        aria-haspopup="true"
        aria-label="account of current user"
        color="inherit"
        size="large"
      >
        <Memo(ForwardRef(AccountCircleIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(Toolbar)>
  </ForwardRef(AppBar)>
  <ForwardRef(SwipeableDrawer)
    onClose={[Function]}
    onOpen={[Function]}
    open={false}
  >
    <ForwardRef(Box)
      sx={
        Object {
          "height": "100%",
          "width": 250,
        }
      }
    >
      <ForwardRef(List)>
        <ForwardRef(ListItem)
          button={true}
          key="Inbox"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Inbox"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Starred"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MailIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Starred"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Send email"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Send email"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Drafts"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MailIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Drafts"
          />
        </ForwardRef(ListItem)>
      </ForwardRef(List)>
      <ForwardRef(Divider) />
      <ForwardRef(List)>
        <ForwardRef(ListItem)
          button={true}
          key="All mail"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="All mail"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Trash"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MailIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Trash"
          />
        </ForwardRef(ListItem)>
        <ForwardRef(ListItem)
          button={true}
          key="Spam"
        >
          <ForwardRef(ListItemIcon)>
            <Memo(ForwardRef(MoveToInboxIcon)) />
          </ForwardRef(ListItemIcon)>
          <ForwardRef(ListItemText)
            primary="Spam"
          />
        </ForwardRef(ListItem)>
      </ForwardRef(List)>
    </ForwardRef(Box)>
  </ForwardRef(SwipeableDrawer)>
</Fragment>
`;
